Mobile/ShopLens/Droid/Activities/ShoppingCartActivity.cs

```diff
@@ -1,50 +1,171 @@
+using System.Text.RegularExpressions;
 
 namespace ShopLens.Droid
 {
     [Activity(Label = "ShoppingCartActivity")]
-    public class ShoppingCartActivity : Activity
+    public class ShoppingCartActivity : Activity, IRecognitionListener, TextToSpeech.IOnInitListener
     {
         readonly string PREFS_NAME = ConfigurationManager.AppSettings["ShopCartPrefs"];
 
         EditText addItemEditText;
         Button addItemButton;
+        Button recogniseVoice;
         ListView listView;
         ActivityPreferences prefs;
 
+        SpeechRecognizer commandRecognizer;
+        Intent speechIntent;
+
         List<string> items;
+        ArrayAdapter<string> listAdapter;
+
+        private TextToSpeech tts;
+        private readonly string voiceCartCmd = ConfigurationManager.AppSettings["CmdVoiceCart"];
 
         protected override void OnCreate(Bundle savedInstanceState)
         {
             base.OnCreate(savedInstanceState);
             SetContentView(Resource.Layout.ShoppingCart);
 
+            commandRecognizer = SpeechRecognizer.CreateSpeechRecognizer(this);
+            commandRecognizer.SetRecognitionListener(this);
+
             prefs = new ActivityPreferences(this, PREFS_NAME);
             items = prefs.GetPreferencesToList();
 
+            tts = new TextToSpeech(this, this);
+
             listView = FindViewById<ListView>(Resource.Id.ShopCartList);
             addItemButton = FindViewById<Button>(Resource.Id.ShopCartAddItemButton);
             addItemEditText = FindViewById<EditText>(Resource.Id.ShopCartAddItemEditText);
+            recogniseVoice = FindViewById<Button>(Resource.Id.ShopCartRecVoice);
 
-            ArrayAdapter<string> listAdapter = 
+            listAdapter = 
                 new ArrayAdapter<string>(this, Android.Resource.Layout.SimpleListItemChecked, items);
             listView.Adapter = listAdapter;
             listView.ChoiceMode = ChoiceMode.Multiple;
 
-            addItemButton.Click += (sender, e) =>
+            addItemButton.Click += AddTextBoxProductToList;
+            recogniseVoice.Click += RecogniseVoice;
+        }
+
+        private void RecogniseVoice(object sender, EventArgs e)
+        {
+            speechIntent = VoiceRecognizerHelper.SetUpVoiceRecognizerIntent();
+            commandRecognizer.StartListening(speechIntent);
+        }
+
+        public void OnInit([GeneratedEnum] OperationResult status)
+        {
+            // If we get an error, use the default language.
+            if (status == OperationResult.Error)
+            {
+                tts.SetLanguage(Locale.Default);
+            }
+            else if (status == OperationResult.Success)
+            {
+                tts.SetLanguage(Locale.Us);
+            }
+        }
+
+        private void SpeakOut(string message)
+        {
+            if (!string.IsNullOrEmpty(message))
+            {
+                tts.Speak(message, QueueMode.Flush, null, null);
+
+                while (tts.IsSpeaking) { }
```

Comment:
We need to wait until the tts has finished speaking. Don't know how to write this differently.